src/maxminddb_stubs.c

```diff
@@ -217,19 +217,23 @@ CAMLprim value mmdb_ml_lookup_path(value ip, value query_list, value mmdb)
 			       (const char *const *const)query);
   check_status(status);
   check_data(entry_data);
-  free(result);
-  for (int i = 0; i < copy_count; free(query[i]), i++);
-  free(query);
+  caml_stat_free(result);
+  for (int i = 0; i < copy_count; caml_stat_free(query[i]), i++);
+  caml_stat_free(query);
   query_r = caml_alloc(2, 0);
   as_mmdb = NULL;
   switch (entry_data.type) {
   case MMDB_DATA_TYPE_BYTES:
     clean_result = caml_stat_alloc(entry_data.data_size + 1);
     memcpy(clean_result, entry_data.bytes, entry_data.data_size);
+    caml_clean_result = caml_copy_string(clean_result);
+    caml_stat_free(clean_result);
     goto string_finish;
 
   case MMDB_DATA_TYPE_UTF8_STRING:
     clean_result = strndup(entry_data.utf8_string, entry_data.data_size);
+    caml_clean_result = caml_copy_string(clean_result);
+    free(clean_result);
```

Comment:
> is there a reason for not calling caml_copy_string directly on the entry_data string?

I guess the reason is that `entry_data` is not `NULL`-terminated.

Suddenly, I see something that resembles a bug: `clean_result` must contain `NULL` at the end (that's why it was allocated with size `data_size + 1` in the first place).

And yes, double allocation is redundant. It would be better to build an empty string with `caml_alloc_string` and copy the contents of `entry_data` there with `memcpy`, while `NULL` will already be at the end of the allocated string.